src/traces/image/calc.js

```diff
@@ -0,0 +1,35 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+// var Registry = require('../../registry');
+// var Lib = require('../../lib');
+var Axes = require('../../plots/cartesian/axes');
+
+module.exports = function calc(gd, trace) {
+    var xa = Axes.getFromId(gd, trace.xaxis || 'x');
+    var ya = Axes.getFromId(gd, trace.yaxis || 'y');
+
+    var x0 = trace.x0 - trace.dx / 2;
+    var y0 = trace.y0 - trace.dy / 2;
+    var h = trace.z.length;
+    var w = trace.z[0].length;
```

Comment:
So we assume `z` to be a 2D arrays with inner arrays all of the same the same length?

Reply:
Yes, that's what this code assumes!

Reply:
Ok, I guess it's fine to only render 2D arrays with inner arrays all of the same the same length, but how do jagged arrays like:

```
var trace = {
  type: 'image',
  z: [
    [[0,0,1], [0,0,1], [0,0,1]], 
    [[0,1,0], [0,1,0]], 
    [[1,0,0]]
}
```

get handled?

Reply:
Right now, it will crash in a fire. Let's handle that properly. 

Whenever a pixel is undefined, we could
1. Draw nothing (which will show whatever is behind)
4. Draw pixel with a color set by the user via a new attribute
2. Draw pixel with all color component set to their minimum

Whenever we hover on an undefined pixel, we could:
1. Not show any hover label
2. Indicated this pixel is not defined
 
I'm tempted to go with 1 and 1. What do you think @etpinard ?

Reply:
I may vote for 1 and 2. cc: https://github.com/plotly/plotly.js/pull/4291
Then for `not show any hover` implementing `hoverongaps: false`.

Reply:
I went with 1 and 1 in 15c06fa1814fa41b2c7a4a60d6d06cd6b90c4a5d but I'm willing to change it to whatever's best :smile_cat:

Reply:
https://github.com/plotly/plotly.js/commit/15c06fa1814fa41b2c7a4a60d6d06cd6b90c4a5d is :ok_hand: - thanks!

We could eventually add `hoverongaps:true` to image trace, but I think a default of `false` makes more sense for image traces.

Reply:
Oh one more thing @antoinerg, instead of

```js
var w = trace.z[0].length;
```

you should use

```js
var maxRowLength = require('../../lib').maxRowLength;
// ...
var w = maxRowLength(trace.z);
```

Reply:
... to handle cases where the max row length isn't in `z[0]`.

Reply:
Done in 2d393ddab83a5492afaf8a2c2690ddd7da3cb15f

Reply:
Great. Resolved!